Feature bjs2 15595

shade2903 opened the pull request:
create CommentController;
create CommentDto;
create CommentMapper;
create CommentService;
create CommentValidator;
create EntityNotFoundException;
create OperationNotAvailableException;
create CommentServiceTest;
create CommentValidatorTest;
create CommentControllerTest;

shade2903 committed:
BJS2-15595 Comment System

create CommentController;
create CommentDto;
create CommentMapper;
create CommentService;
create CommentValidator;
create EntityNotFoundException;
create OperationNotAvailableException;

shade2903 committed:
BJS2-15595

create CommentServiceTest;
create CommentValidatorTest;
create CommentControllerTest;

ravshanbekn commented on src/main/java/faang/school/postservice/exception/EntityNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class EntityNotFoundException extends RuntimeException {
+    public EntityNotFoundException(String message) {
+        super(message);
```
По моему, такой exception есть в 
jakarta.persistence.EntityNotFoundException;

shade2903 replied:
Ок.  Согласен, если такой есть лучше его переиспользовать. Заменил на jakarta.persistence.EntityNotFoundException;

ravshanbekn commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,69 @@
+        commentValidator.existComment(commentDto.getId());
+        Comment comment = commentRepository.findById(commentDto.getId()).orElseThrow();
+        commentValidator.checkUserIsAuthorComment(comment, userContext.getUserId());
+        comment.setUpdatedAt(LocalDateTime.now());
```
По моему, hibernate автоматически обновляет

shade2903 replied:
Согласен. Просто хотел перестраховатся. Сейчас увидел аннотацию на сущности, да это лишняя операция. Убрал.

ravshanbekn commented on src/main/java/faang/school/postservice/controller/CommentController.java:
```diff
@@ -0,0 +1,37 @@
+    @GetMapping("/posts/{id}")
+    public void findAllByPostId(@Positive @PathVariable long id){
+        commentService.findAllByPostId(id);
+    }
```
@ ResponseStatus был бы уместен, я считаю

shade2903 replied:
Ок. добавил

ravshanbekn reviewed (commented):
Как по мне, написано хорошо

shade2903 committed:
BJS2-15595 Comment System

add ResponseStatus in CommentController
use EntityNotFoundException from jakarta.persistence.EntityNotFoundException;
remove custom EntityNotFoundException;
fix bag in CommentService in method updateComment();

pablo023 commented on src/main/java/faang/school/postservice/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,21 @@
+    @Mapping(source = "postId", target = "post.id")
+    Comment dtoToEntity(CommentDto commentDto);
+
+    List<CommentDto> entitiesToDtos(List<Comment> projectList);
```
А тут он подхватит логику сеттинга айдишек?

shade2903 replied:
Добавил @Mapping(source = "post.id", target = "postId")

pablo023 commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,68 @@
+    @Transactional
+    public CommentDto createComment(CommentDto commentDto) {
+        commentDto.setAuthorId(userContext.getUserId());
+        commentValidator.existUser(commentDto.getId());
```
Ты тут вытаскиваешь айдишник не пользователя, а из дтошки тащишь просто айди

shade2903 replied:
Поправил.

pablo023 commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,68 @@
+    private final CommentValidator commentValidator;
+
+    @Transactional
+    public CommentDto createComment(CommentDto commentDto) {
+        commentDto.setAuthorId(userContext.getUserId());
+        commentValidator.existUser(commentDto.getId());
+        commentValidator.existPost(commentDto.getPostId());
+        Comment comment = commentMapper.dtoToEntity(commentDto);
+        return commentMapper.entityToDto(commentRepository.save(comment));
+    }
```
Надо бы тут пост тащить из базы, ему в список комментов добавлять этот коммент, и сохранять в базу.

shade2903 replied:
Исправил.

pablo023 commented on src/main/java/faang/school/postservice/service/CommentService.java:
```diff
@@ -0,0 +1,68 @@
+            log.error(String.format(msg, postId));
+            throw new EntityNotFoundException(String.format(msg, postId));
+        }
+        return comments.stream()
+                .sorted((commentOne, commentTwo) -> commentTwo.getCreatedAt().compareTo(commentOne.getCreatedAt()))
+                .map(commentMapper::entityToDto)
+                .toList();
+    }
```
return comments.stream()
               .sorted(Comparator.comparing(Comment::getCreatedAt).reversed())
               .map(commentMapper::entityToDto)
               .collect(Collectors.toList());

shade2903 replied:
Поправил, думал может так быстрее будет.

pablo023 reviewed (changes requested):
Надо поправить чутка

shade2903 committed:
BJS2-15595

fix bags in CommentService;
fix bag in CommentMapper;

shade2903 reviewed (commented):
подправил баги.
